servo/connectors/newrelic.py

```diff
@@ -0,0 +1,362 @@
+    #             result = response.json()
+    #             return f"returned {len(result)} results"
+
+    #     return self.config.metrics, query_for_metric
+
+@metadata(
+    description="Newrelic Connector for Opsani",
+    version="0.0.1",
+    homepage="https://github.com/opsani/newrelic-connector",
+    license=License.APACHE2,
+    maturity=Maturity.EXPERIMENTAL,
+)
+class NewrelicConnector(BaseConnector):
+    """NewrelicConnector objects enable servo assemblies to capture
+    measurements from the [Newrelic](https://newrelic.com/) metrics server.
+    """
+    config: NewrelicConfiguration
+
+    # TODO
+    # @on_event()
+    # async def check(self,
+    #     filter_: Optional[Filter] = None, 
+    #     halt_on: HaltOnFailed = HaltOnFailed.requirement
+    # ) -> List[Check]:
+    #     """Checks that the configuration is valid and the connector can capture        
+    #     measurements from Newrelic.
+
+    #     Checks are implemented in the NewrelicChecks class.
+
+    #     Args:
+    #         filter_ (Optional[Filter], optional): A filter for limiting the
+    #             checks that are run. Defaults to None.
+    #         halt_on (HaltOnFailed, optional): When to halt running checks.
+    #             Defaults to HaltOnFailed.requirement.
+
+    #     Returns:
+    #         List[Check]: A list of check objects that report the outcomes of the            
+    #             checks that were run.
+    #     """        
+    #     return await NewrelicChecks.run(self.config, filter_, halt_on=halt_on)
+
+    @on_event()
+    def describe(self) -> Description:
+        """Describes the current state of Metrics measured by querying Newrelic.
+
+        Returns:
+            Description: An object describing the current state of metrics
+                queried from Newrelic.
+        """
+        return Description(metrics=self.config.metrics)
+
+    @property
+    @on_event()
+    def metrics(self) -> List[Metric]:
+        """Returns the list of Metrics measured through Newrelic queries.
+
+        Returns:
+            List[Metric]: The list of metrics to be queried.
+        """
+        return self.config.metrics
+
+    @on_event()
+    def instances(self) -> List[str]:
+        async with self.api_client as client:
+            try:
+                response = await client.get('/config')
+                response.raise_for_status()
+            except (httpx.HTTPError, httpcore._exceptions.ReadTimeout, httpcore._exceptions.ConnectError) as error:                
+                self.logger.trace(f"HTTP error encountered during oco config GET {self.optimizer.api_url}/config: {error}")
+                raise
+
+        opsani_config = response.json()
+        canary = opsani_config.get('adjustment', {}).get('control', {}).get('userdata', {}).get('canary')
+        canary_ip = canary.get('ip')
+        canary_ip = re.sub(r"\.", "-", canary_ip)
+        canary_hostname = f"ip-{canary_ip}.ec2.internal"
```

Comment:
This substitution only works with a specific customer.  In my more generic testing, these hostnames look more like this: `canary_hostname = f"ip-{canary_ip}.us-west-2.compute.internal"`

This will likely be different per customer, unless they are using AWS defaults, in which case they'll look like my example with the exception of the respective region.  We'll want to make this more dynamic and/or configurable with sane defaults

Reply:
this whole canary hostname computation feels brittle and inelegant to me, I think we can do better. But in any case, the logic definitely does not belong in the middle of the New Relic connector

Reply:
Moved the format string to the configuration: https://github.com/opsani/servox/blob/279ae1fbd2709246baf68615e400c596adf60623/servo/connectors/newrelic.py#L94

Reply:
We should evict this out into a Harness specific connector. We don't want our New Relic connector carrying around implementation details about taking private IP addresses from EC2 instances running in Fargate that have been posted to a private API and are then interpolated into this hostname template which is tightly coupled to internal EC2 instances running in the `us-west-2` data center/availability zone.

We have a couple of distinct but intersecting concerns that are getting gummed up here